---
title: "Aza_Toca_Pablo-PEC2"
author: "Pablo Aza"
date: "2023-12-21"
output: 
 pdf_document:
    toc: true
    toc_depth: 2
    df_print: kable
    

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1.Introduction

Microarray technologies have revolutionized genomic research by allowing the simultaneous measurement of thousands of genes' expression. These devices provide a comprehensive view of gene activity, offering valuable data to understand the biological processes underlying normal and pathological conditions.

However, the complexity and magnitude of the datasets generated by microarrays pose significant challenges in terms of analysis and extraction of relevant information. This is why the use of bioinformatics tools is becoming increasingly important to analyze such data. In this context, the Bioconductor project offers a set of tools in R designed for the analysis of high-dimensional biological data, such as those generated by microarrays.

In this work, we will leverage the potential of Bioconductor to explore a microarray from initial data exploration steps, through differential expression analysis, to identifying overrepresented biological processes.

The selected dataset is "GDS1736" from the "GSE3737" series of Rattus norvegicus in the study "Acute hypotension effect on kidneys." The study involves profiling the gene expression of kidneys from 12-week-old normotensive Wistar-Kyoto mice subjected to acute hypotension by withdrawing 3 ml of blood. The results provide insights into vasoregulatory mechanisms and the genetics and physiology of blood pressure regulation.


# 2. Objectives

The main objective is to analyze the differential gene expression between the Control (normotensive) and Hypotensive Mouse samples. Additionally, we will study, through enrichment analysis, which biological processes may have been affected.

# 3. Materials and Methods

## 3.1 Data Acquisition

The data is downloaded from the GEO database using R and the GEOquery package. The series number used as a query is GSE2401.

## 3.2 Microarray Exploration

General exploration is performed using BoxPlot and hierarchical clustering dendrograms. The arrayQualityMetrics() function and basic R functions are used for this purpose. We conduct a Principal Component Analysis using the prcomp() function and visualize the data using the ggplot() package.

## 3.3. Differential Analysis

Differential expression analysis of genes is performed using the Limma package. 

## 3.4 Enrichment Analysis

We use the rat2302.db package from the Affymetrix Rat Genome 230 2.0 Array microarray platform. The clusterProfiler library is employed for enrichment analysis.

# 4. Results

## 4.1 Data Download

o obtain the data, we use GEOquery to access the public GEO database for biological experiments. With GEOquery, we can download this data directly from R. The selected GEO Dataset is GDS1251.

- GEO Dataset: GDS1251
- GEO Serie: GSE2401
- Platform: GPL341 

```{r, warning=FALSE, message=FALSE}


# Install all necessary packages for analysis

if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

if (!requireNamespace("ggplot2", quietly = TRUE)) {
  install.packages("ggplot2")
}


# List of packages
packages <- c("GEOquery","limma","rat2302.db","clusterProfiler")

# Install packages
missing_packages <- packages[!(packages %in% installed.packages()[,"Package"])]
if(length(missing_packages) > 0) {
  BiocManager::install(pkgs = missing_packages)
}


require(GEOquery)
require(limma)
require(rat2302.db)
require(clusterProfiler)

#Abrimos el dataset
gds<-getGEO("GDS1251")

```

Once we have initial information about the data, we generate the content matrix using the dataset number and creating a GDS file.

With the Meta() function, we extract information about the experiment.

```{r, warning=FALSE, message=FALSE}
Meta(gds)
```

We have a description of the study: "Expression profiling of kidneys of 12-week-old normotensive Wistar-Kyoto males subjected to acute hypotension by withdrawing 3 ml of blood. Results provide insight into vasoregulatory mechanisms and the genetics and physiology of blood pressure regulation."

We also observe the two groups that compose it:

-Control
-Hypotense

The next step is to extract the expression set.

```{r, warning=FALSE, message=FALSE}
#extraer el expresionset y eliminar valores NA
matrix.data<-GDS2eSet(gds,do.log2 = TRUE)
matrix.data<-na.omit(matrix.data)
print(matrix.data)

```


## 4.2 Exploration, Quality Control

For exploration and quality control, we can use the "arrayQualityMetrics(raw_data, outdir = "Quality_Report_Kidney")" package. However, for convenience, we will perform data exploration with "ad-hoc" functions.

Use a boxplot to analyze the intensity distribution.

```{r, warning=FALSE, message=FALSE}
boxplot(exprs(matrix.data),names=pData(matrix.data)[,2], which="all", las = 2, main = "Intensity distribution of RAW data",
    cex.axis = 0.6)
```

We observe that the distribution of signals is very similar between different samples. It is noteworthy that this is expected since the data was pre-normalized.

Next, we create a hierarchical clustering dendrogram to see how our microarrays would group.

```{r, warning=FALSE, message=FALSE}
# HIERARQUICAL CLUSTERING
clust.euclid.average <- hclust(dist(t(exprs(matrix.data))), method = "average")
plot(clust.euclid.average, labels = pData(matrix.data)[,2], main = "Hierarchical clustering of RawData",
    cex = 0.7, hang = -1)
```

The Cluster analysis reveals that they do not group as expected.

We perform a Principal Component Analysis (PCA) to see if we can simplify the data dimensions while maintaining the same variability. In our case, we consider genes as variables, so we transpose the dataframe. If we don't, we would see how genes are related to each other, which is not what we want.

Since we have 9 samples in our analysis, we will see 9 PCs when we represent them. When we do, we will see that the first PC has the most variation in the original data, and so on.

```{r, warning=FALSE, message=FALSE}
# Find infinite or missing values that cause problems in PCA
any(is.na(exprs(matrix.data)))
datos<-na.omit(exprs(matrix.data))

# Principal Component Analysis
pca<-prcomp(t(datos), scale=TRUE)

# Prepare data for ggplot
pca.data<-data.frame(Sample=pData(matrix.data)$stress,X=pca$x[,1],Y=pca$x[,2])


# Plot
library(ggplot2)

pca.var<-pca$sdev^2 #sacar desviacion estandar
pca.var.tot<-round(pca.var/sum(pca.var)*100,1) #los porcentages de DesEst explicados para cada PC

ggplot(data=pca.data, aes(x=X, y=Y, label=Sample))+ geom_text()+
  xlab(paste("PCA1-",pca.var.tot[1],"%",sep=" "))+
  ylab(paste("PCA2-",pca.var.tot[2],"%",sep=" "))+
  theme_bw()+
  ggtitle("PCA Graph")
```

The axes show the percentage of variability in the original data explained. As observed, we do not see the two clusters well differentiated as expected because controls and cases are intermingled.

PCA and dendrogram analysis could indicate a technical or experimental error, the day of sample collection, or their mixing. It could also be due to significant biological variability within each group, making it challenging to identify distinctive patterns in PCA. Additionally, there might be subtypes within the groups that were not considered in the study (e.g., gender).

Despite the above, we continue with the next section regarding the selection of differentially expressed genes.

## 4.3 Selection of Differentially Expressed Genes

To select differentially expressed genes, considering that we have a single factor with 2 levels (Control/Hypotension), we could use t-tests and select genes based on their adjusted p-value. However, we choose to use linear models for gene selection, given the activity's statement.

We formulate the specific question to be answered:

*Which genes are differentially expressed between control and Hypotensive mice?*

We display the model parameterization; a two-level factor.

\(Yi,j = \alpha i  + Ei,j\)

Where  i=1,2 because there can be up to 2 levels, and  j=1, ...5 since there can be up to 5 biological replicates. 

We set up the design matrix and the linear model of the study. In our case, we have 5 controls and 4 Hypotension:


\[
\left(\begin{array}{cc} 
Y1\\
Y2\\
Y3\\
Y4\\
Y5\\
Y6\\
Y7\\
Y8\\
Y9\\
\end{array}\right)
=
\left(\begin{array}{cc} 
1 & 0\\
1 & 0\\
1 & 0\\
1 & 0\\
1 & 0\\
0 & 1\\
0 & 1\\
0 & 1\\
0 & 1\\
\end{array}\right)
*
\left(\begin{array}{cc} 
\alpha1\\ 
\alpha2\\
\end{array}\right)
+
\left(\begin{array}{cc} 
\epsilon 1\\
\epsilon 2\\
\epsilon 3\\
\epsilon 4\\
\epsilon 5\\
\epsilon 6\\
\epsilon 7\\
\epsilon 8\\
\epsilon 9\\
\end{array}\right)
\]

The contrast we are considering would be:

\(\beta = \alpha 1  - \alpha 2\)

With the following contrast matrix:

\[
\left(\begin{array}{cc} 
\beta1\\
\end{array}\right)
=
\left(\begin{array}{cc} 
1 & -1\\
\end{array}\right)
*
\left(\begin{array}{cc} 
\alpha1\\ 
\alpha2\\
\end{array}\right)
\]

Once we have the formulation, we proceed to estimate the parameters and make the comparison. We use the limma package for this purpose.

The following code defines the design matrix and the contrasts.

```{r, warning=FALSE, message=FALSE}

#establecemos los niveles
lev<-factor(pData(matrix.data)$stress)

#diseñamos la matriz diseño
design<-model.matrix(~0+lev)
rownames(design)<-pData(matrix.data)$stress
colnames(design)<-c("hypo","control")
print(design)

#elaboramos la matriz de contraste
cont.matrix<-makeContrasts(Hypo.Control=(control-hypo),levels=design)
cont.matrix
```

In the next step, we calculate the contrasts.

```{r, warning=FALSE, message=FALSE}
fit<-lmFit(matrix.data,design)#modelo linea
fit.main<-contrasts.fit(fit, cont.matrix)#contraste
fit.main<-eBayes(fit.main)#regularizar varianza
```

As a result, we have the Fold-changes and adjusted p-values, which allow us to design the "Volcano-plot" to visualize the data. It is worth noting that by using Bayes regularization, we are improving the estimation of gene variability, acting as "shrinkage," especially when the number of samples is small, as in our case: 5 for control and 4 for Hypotension.

Next, we visualize the data with a volcano plot and additionally, using the topTable function, we generate a list of genes that are differentially expressed for the contrast.

```{r, warning=FALSE, message=FALSE}
coefnum = 1
opt <- par(cex.lab = 0.7)
volcanoplot(fit.main, coef=coefnum, highlight=10, names=fit.main$ID,
main=paste("Differentially expressed genes",
colnames(cont.matrix)[coefnum], sep="\n"))
abline(v=c(-1,1))
par(opt)


#ordered data list
Hypo.Control <- topTable (fit.main, number=nrow(fit.main), coef="Hypo.Control", adjust="fdr")
```

## 4.4 Results Annotation.

As we are working with the genome of Rattus norvegicus, we searched for the reference "Annotation Packages" on Bioconductor. After searching, it seems that the appropriate package is the Affymetrix Rat230_2 Array annotation data, named "rat2302.db."

```{r, warning=FALSE, message=FALSE}

keytypes(rat2302.db)
```
Next, we annotate the data using identifiers such as "Symbol," "EntrezID," and "EnsemblID."

-SYMBOL: A readable identifier.
-ENTREZID: A unique numerical identifier assigned to each gene by the National Center for Biotechnology Information (NCBI) through its Entrez Gene database.
-ENSEMBL: A unique numerical identifier assigned to each gene by the National Center for Biotechnology Information (NCBI) through its Entrez Gene database.

```{r, warning=FALSE, message=FALSE}

geneAnots <- AnnotationDbi::select(rat2302.db,rownames(Hypo.Control) , c("SYMBOL", "ENTREZID", "ENSEMBL"))
head(geneAnots)
```

We perform an enrichment analysis, which is used to identify gene classes overrepresented in a set of interest (hypotensive mice) compared to a background set (controls).

We will conduct an "Over-Representation Analysis" using the "clusterProfiler" package. First, we select genes that are considered differentially expressed, i.e., those with a P.value less than 0.05 and a logFC greater than 1.

```{r, warning=FALSE, message=FALSE}
select.exp<-Hypo.Control[Hypo.Control$P.Value < 0.05 & Hypo.Control$logFC > 1, ]
dim(select.exp)

```

Once we have the genes considered differentially expressed (a total of 212), we perform the enrichment analysis on the resulting dataset. It's worth noting that we will use "BP" (Biological Process) as the sub-ontology to focus on higher-level biological processes involving differentially expressed genes.

```{r, warning=FALSE, message=FALSE}
#Extract the ENTREZID of overexpressed genes, which is an argument in the function as a vector

entrez.select.exp <- AnnotationDbi::select(rat2302.db,rownames(select.exp) , c("ENTREZID"))

#Remove duplicates and NA values
entrez.select.exp<-na.omit(entrez.select.exp)



ego <- enrichGO(gene = as.integer(entrez.select.exp$ENTREZID), #a vector of entrez gene id
                keyType = "ENTREZID",
                OrgDb = rat2302.db, 
                ont = "BP", #Subontology
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.25, 
                readable = TRUE)
```

Export the results for access.

```{r, warning=FALSE, message=FALSE}
# Get the current working directory
directorio <- getwd()

# Create "results" directory if it doesn't exist
if (!file.exists("results")) {
  dir.create("results")
}

# Define the path for the results directory
resultsDir <- file.path(directorio, "results")

# Set working directory
setwd(resultsDir)

# Data.frame of Enrichment Analysis
ego_results <- data.frame(ego)
write.csv(ego_results, "clusterProfiler_ORAresults_UpGO.csv")

setwd(directorio)
```

Let's visualize the enrichment analysis results to see which processes are more overexpressed.

- Dotplot of terms: 

```{r, warning=FALSE, message=FALSE}
dotplot(ego, showCategory=10)
```

In this Dotplot, we can see enriched biological processes, their significance, and the ratio of differentially expressed genes (X-axis) associated with these processes. In this case, processes related to the muscular system are prominent.


-Visualization of GO terms in hierarchy:

```{r, warning=FALSE, message=FALSE}
goplot(ego, showCategory=3)
```

Here, we explore the hierarchical structure of enriched terms. We focus on higher-level terms such as "Muscle System Process" and also observe "Cell-Substrate Adhesion."

- Enrichment Map

```{r, warning=FALSE, message=FALSE}
## Enrichmap clusters the 50 most significant (by adj.P.Va) GO terms to visualize relationships between terms
library(enrichplot)
ego_sim <- pairwise_termsim(ego)
emapplot(ego_sim, cex_label_category=0.5)
```

### 4.4.2 Gene Set Enrichment Analysis (GSEA)

Unlike the previous analysis, GSEA does not focus on the differential expression of genes. This method is used to interpret gene expression data and identify whether predefined sets of genes show statistically significant differences in their expression levels between our two conditions. In summary, we obtain information about biological pathways associated with observed changes in gene expression.

```{r, warning=FALSE, message=FALSE}

#Add ENTREZID to all our genes
entrezIDs <- AnnotationDbi::select(rat2302.db, rownames(Hypo.Control), c("ENTREZID"))


# Add ENTREZID to our genes by merging into the final geneList object
probeid.select.exp<- cbind( PROBEID= rownames(Hypo.Control), Hypo.Control)
geneList  <- merge(probeid.select.exp, entrezIDs, by="PROBEID")

# Remove genes with small logFC
geneList <- geneList[order(abs(geneList$logFC), decreasing=T),]
geneList <- geneList[ !duplicated(geneList$ENTREZ), ]  ### Keep highest

# Reorder for GSEA analysis
geneList <- geneList[order(geneList$logFC, decreasing=T),]
genesVector <- geneList$logFC
names(genesVector) <- geneList$ENTREZ


gseResulti <- gseKEGG(geneList = genesVector,
                      organism = "rno",
                      keyType = "kegg",
                      exponent = 1,
                      minGSSize = 10,maxGSSize = 500,
                      pvalueCutoff = 0.05,pAdjustMethod = "BH",
                      verbose = TRUE,
                      use_internal_data = FALSE,
                      seed = TRUE,
                      eps=0,
                      by = "fgsea"
                )

# keggResultsList[[i]] <- gseResulti
```


```{r, warning=FALSE, message=FALSE}
library(kableExtra)
gsea.result <- setReadable(gseResulti, OrgDb = rat2302.db, keyType ="ENTREZID" )

gsea.result.df <- as.data.frame(gsea.result)

gsea.result.df

print(kable(gsea.result.df[,c("Description","setSize","NES","p.adjust")])%>% scroll_box(height = "500px"))
```

The main difference between GSEA and ORA is that the latter compares the overlap between two gene sets (differential genes of control and hypotensive mice), using statistical tests to determine significance. GSEA, on the other hand, classifies all genes in the dataset from which we started and then checks if genes in a gene set (associated with a process) are enriched. While ORA focuses on the overlap between sets of genes that have been verified as differential, GSEA considers the entire distribution of changes in gene expression. In summary, GSEA considers the complete distribution of changes in gene expression in a dataset.

# 5. Discussion

We are dealing with a study involving two groups of mice: control and those with induced hypotension. Our analyses of differential gene expression, supported by enrichment analysis results, suggest that certain biological processes such as cell adhesion or the muscular system show differences. The association of differentially expressed genes with processes of the muscular system and cell-substrate adhesion could have interesting implications in the context of hypotension according to our results.

Muscle function, especially contractility and response to blood pressure, is essential for blood pressure regulation. Therefore, it is logical for hypotensive mice to have higher expression of these genes because they need an adaptation of the muscular system to changes in blood pressure. Furthermore, cell-substrate adhesion is crucial for the integrity and function of tissues, such as muscles. Therefore, these changes could reflect cellular adaptations to hypotension by hypotensive mice.

Additionally, the ketone response systems are also active in hypotensive individuals. This could be associated with the use of ketone bodies as an energy source. This finding suggests a metabolic reprogramming in response to hypotension. Changes in energy metabolism could be an adaptive strategy to maintain homeostasis under hypotensive conditions.
 

# 6. Conclusion

In summary, ORA results suggest that induced hypotension is associated with changes in gene expression affecting muscle function, cell adhesion, and metabolic response.